Added Dockerfile for a Docker deployment

adundovi opened the pull request:
This one is based on Alpine Linux.

One needs to set-up an external mongodb (or use existing one), edit the pump.io.json config and run the build command. In my case it worked.

strugee commented:
@adundovi fantastic! I'll try to look at this soon, although I'm not sure when I'll be able to find time.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+  make \
+  g++ \
+  graphicsmagick
+RUN npm install -g databank-mongodb
```
Use `databank-mongodb@0.19.2`; see #1287

adundovi replied:
OK.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM alpine:3.5
+
+LABEL maintainer="adundovi@gmail.com"
+LABEL description="Prepares a docker image \
```
Pretty sure this should be a description of the built image itself, so please replace with `Stream server that does most of what people really want from a social network.`

We should probably change that to something less jargon-y, to be honest. But in the meantime...

adundovi replied:
OK. I didn't pay attention to these lines and of course that everything which has some reasons behind is a better choice. Also, if you have an official pump.io e-mail address, we can put it there.

strugee replied:
We don't, yours is fine for now

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+  g++ \
+  graphicsmagick
+RUN npm install -g databank-mongodb
+RUN npm install -g pump.io
```
I think it would be better to find a way to have this install from the source tree instead of from npm. What do you think?

adundovi replied:
I'm unfamiliar with node.js waters, therefore I don't know when and how often do you push something to npm, but I generally avoid the source tree in something which is intended for "just work" environment because it could be unstable or even undeployable while one wouldn't push something in that state in a package manager, right? Or you suggest pulling the source from a tagged release?

strugee replied:
@adundovi the images we'd upload to Docker Hub would be built from tagged releases. Building the image from the cloned source tree is good for a couple reasons:

1. If you check out a tag and build the Docker image, you'll always get an image with the code at that tag (instead of whatever's the latest version on npm)
2. We can test that the Docker image works in CI
3. It'll work for development, for contributors who already have Docker set up and find it convenient to use it to manage the environment

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+    "ownerURL":  "http://dockerio.cloudapp.net",
+    "hostname":  "pumpio.example.com",
+    "port":  443,
+    "urlPort": 443,
```
Redundant since we're already binding directly to port 443 (this option exists only to support reverse-proxy setups)

adundovi replied:
I know it is redundant, but I personally use this option because I deploy pump.io behind the nginx proxy so I would like to make visible all available options in this generic config sample.

strugee replied:
Unless you're running nginx inside the Docker container, it won't matter. The reason `urlPort` exists is so that you can specify `443` in URLs but still bind to some high-numbered port. The only reason you'd need to bind to a different port is because something (i.e. nginx) is already using port 443 on the system's primary network interface.

This isn't relevant in the Docker case since the container will get its own private network with its own private port 443. So `urlPort` doesn't matter no matter what.

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "driver":  "mongodb",
+    "params":  {"dbname": "pumpio", "host":"127.0.0.1", "port":27017},
+    "secret":  "put a secret string here",
```
Aren't all these config values supposed to be configurable by whoever's installing the Docker image?

adundovi replied:
Values here will be hard-coded in the image itself, so one needs to modify them before building the image, but there is an option to override the whole path with the external file on the host (similar like with the certificates below) as far as I know. I must admit that I'm not a docker guru, but I haven't seen quite different approaches in the official docker library ( https://hub.docker.com/u/library/ ). If you know examples how to make an image more flexible and configurable without editing config files, I'm listening.

strugee replied:
I don't know offhand, unfortunately. Telling admins to just bring their own pump.io.json seems relatively reasonable to me...

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "driver":  "mongodb",
+    "params":  {"dbname": "pumpio", "host":"127.0.0.1", "port":27017},
```
This is pointing to a MongoDB running on localhost, but the admin is supposed to run MongoDB outside containers based on this image. This won't work, AFAICT.

adundovi replied:
Yes, that's true, but I thought it will be changed in any case. Maybe putting x.x.x.x instead?

strugee replied:
Ah yeah, we should make this configurable anyway. You're right.

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+    "port":  443,
+    "urlPort": 443,
+    "key": "/etc/ssl/private/privkey.pem",
+    "cert": "/etc/ssl/private/fullchain.pem",
```
How are these supposed to get into the image?

adundovi replied:
Either by copying or by mounting from the host during the run. I'm more in favour of the second option...

strugee replied:
Agreed. Seems like we'd _have_ to do it that way to publish on Docker Hub.

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+    "key": "/etc/ssl/private/privkey.pem",
+    "cert": "/etc/ssl/private/fullchain.pem",
+    "address":  "0.0.0.0",
+    "nologger":  false,
```
This turns on logging but doesn't specify a logfile. I can't remember if that's a problem? Aren't Docker logs supposed to go to stdout?

adundovi replied:
Pump.io in this case write logs in stderr and stdio, right? This is recommended because Docker will show that in its logs. See: https://docs.docker.com/engine/reference/commandline/logs/#parent-command

strugee replied:
Oh yep, you're right.

strugee commented on deploy/docker/pump.io.json:
```diff
@@ -0,0 +1,23 @@
+    "address":  "0.0.0.0",
+    "nologger":  false,
+    "serverUser":  "pumpio",
+    "datadir": "/data/uploads",
```
Putting `uploads` at the end of this path is pretty confusing - this will result in all uploaded files going into `/data/uploads/uploads`. I think a good place to put this is `/var/local/pump.io`, which is in line with the Filesystem Hierarchy Standard

adundovi replied:
The datadir needs to be persistent, right? That is the reason why I would put it in an unusual location. People tend to think that everything what is in the standard hierarchy will be part of the image and thus destroyed together with the image. But I don't have a strong opinion on that.

strugee replied:
Ah, okay. That sounds fine.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN mkdir -p /etc/ssl/private/
+
+VOLUME "/data/uploads"
+EXPOSE  443
```
Needs to expose port 80 too, for the bounce server

adundovi replied:
OK.

strugee commented on deploy/docker/README.md:
```diff
@@ -0,0 +1,18 @@
+######################
```
Remove this line, please

strugee commented on deploy/docker/README.md:
```diff
@@ -0,0 +1,18 @@
+######################
+
+### pump.io http://pump.io Docker image - listens port 443
```
Remove the URL and the "listens port 443" bit

strugee commented on deploy/docker/README.md:
```diff
@@ -0,0 +1,18 @@
+
+### pump.io http://pump.io Docker image - listens port 443
+
+1) Copy pump.io.json.sample to pump.io.json and configure pump.io.json according to the needs
```
Ultimately we want to publish a reusable image to Docker Hub so I don't think this is the way to go. We need to make stuff configurable at runtime. (Unless what I'm thinking of isn't actually possible? 95% sure it is though)

adundovi replied:
I understand what you want... I'll redo the dockerfile with env variables - that is the only alternative option that I am aware of to enable something like that. For example ` --env PUMPIO_DBNAME="something"`.

strugee replied:
Hmm, okay. pump.io doesn't currently accept configuration through env variables but there are plans to... I can reprioritize those. See #1202

strugee commented:
Still haven't run this locally but that's probably enough for now :)

Thanks so much for taking this on! I really appreciate it :heart:

adundovi commented:
OK. I have fixed all the issues and introduced a possibility of configuring the container before starting it, but after building it using ENV variables. Basically that's it.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -1,28 +1,44 @@
 
 ENV LC_ALL en_US.UTF-8
 
+ENV PUMPIO_VERSION 3.0.2
```
Is there any way we can avoid hardcoding this? I.e. is there any way we could shell out to extract this version from `package.json`?

strugee replied:
Responding to https://github.com/pump-io/pump.io/pull/1307#discussion_r109278471 - the point of maintaining it in the same repository is that we can update the Dockerfile at the same time we introduce new dependencies. So your second point isn't relevant, and we'd already have a cloned version of `package.json`.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -1,28 +1,44 @@
+    g++ \
+    git \
+    graphicsmagick \
+ && adduser -D -g "" pumpio \
```
Will failed `RUN` commands not stop the Docker build? If not, I think it looks nicer to just replace all these `&&`s with more `RUN` commands

strugee replied:
@adundovi was this intended to be a reply to https://github.com/pump-io/pump.io/pull/1307#discussion_r109271410?

adundovi replied:
Yes, sorry. Regarding the one RUN command - it creates less layers which is [desirable](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#minimize-the-number-of-layers).

strugee replied:
Oh, each `RUN` command creates new layers? Didn't realize that. Makes sense to keep it as-is then.

strugee commented on deploy/docker/Dockerfile:
```diff
@@ -1,28 +1,44 @@
+ && mkdir -p /etc/ssl/private \
+ && npm install -g databank-mongodb@0.19.2 \
+ && mkdir -p /usr/src \
+ && curl -fSL https://github.com/pump-io/pump.io/archive/v$PUMPIO_VERSION.zip -o pumpio.zip \
```
Again, I'd prefer for this to be built from the currently checked-out tree. That lets people use Docker for development, plus downloading archives from GitHub when you already have a local copy feels kind of sketchy.

(That and the fact that this will probably break my release workflow - ideally we want to automatically publish Docker images whenever a new release comes out, which means automatically doing it during `npm version` or `npm publish` or something, when I might not have pushed the tag yet.)

adundovi replied:
Hmm, but this part I don't understand. If I would develop something, I would never do it through a docker image because it would slow down my workflow: rebuilding the image every time when I want to test something doesn't make sense. Docker is primary here for deployment. One can do the development through some other techniques, like LXC. A prescription how to set up the developing environment is already here, in this Dockerfile.

Also, it would make sense to decouple the docker image creation from the original source, ideally to have a separate repository for it which means that the local copy is not available anymore.

Building a docker image shouldn't depend on something outside of the folder where the Dockerimage file is located (it is against the rules). Either we would put this Dockerimage inside the root folder of this repository (../..), or we keep it like this. A script can be made which would update the dockerfile once you publish a new version with publish, right?

Generally, I don't agree with this comment, but it seems that the disagreement arises due to different visions for what Docker will be used. We should clarify it. For me it is used for the stable deployment of Pump.io and for you it some sort of development env, right?

strugee replied:
Yeah, I personally find the no-Docker workflow just fine but apparently a lot of people prefer to develop in Docker containers. The logic being that you set up Docker once and then don't worry about setting up a development environment again, regardless of language, dependencies, etc. See https://medium.com/iron-io-blog/why-and-how-to-use-docker-for-development-a156c1de3b24, https://docs.docker.com/opensource/project/set-up-dev-env/, etc.

However it looks like those projects don't have you rebuild the image every time... I'm not familiar with Docker enough to understand exactly how it works :(

> Also, it would make sense to decouple the docker image creation from the original source, ideally to have a separate repository for it which means that the local copy is not available anymore.

This doesn't make sense to me. I'd like to track the Docker build in the same repo so when we e.g. introduce a new dependency, we can update the Dockerfile in lockstep. Giving the Docker build a separate history makes it difficult to see that correlation. And as far as I can tell, there are no disadvantages to this approach. Am I missing something?

> Building a docker image shouldn't depend on something outside of the folder where the Dockerimage file is located (it is against the rules).

This seems to be an argument in my favor. Wouldn't your `curl` command violate that rule? :)

> A script can be made which would update the dockerfile once you publish a new version with publish, right?

Yes, but if possible I'd prefer to avoid this. Less moving parts. If it's not technically feasible though, a script can be written.

adundovi replied:
I have read your links, and I now understand this use case, but I must say it is a totally different approach compared to a deployment container because a source code is started from the outside of the container (that's why there are no rebuilds). Therefore, I recommend that we split the docker story to deployment and development one. For the former one we start a new repository which would be updated only when needed (ideally, rarely) and the later which would follow the practice from the first link you submitted and where the dockerfile would be in the root directory of this repository.

For the deployment one - I don't want to manually checkout a github repository, maybe I don't even have git installed on my host. I just want to have a pump.io instance running by using docker pull command.

strugee replied:
Yeah, I'm starting to think that we might want to have two Dockerfiles too. I'll think about how I'd like to do that and get back to you. (To be clear though, I was never saying that you should have to have git installed to use the deployment one. I was just saying that you'd have to have git and a local clone if you wanted to _build_ the deployment image.)

strugee replied:
> For the former one we start a new repository which would be updated only when needed (ideally, rarely)

So I think this is the only controversial point here - the rest we're generally in agreement about. I'd like to again state that I don't see any compelling advantages to having a separate repository for the production Dockerfile, whereas I see advantages to keeping it in the same repo. What exactly is the problem with that? I know Dockerfiles are supposed to only rely on stuff in the directory they're in (where is that recommendation from, btw?) but I don't really think that's a compelling argument. Almost no one is going to download the Dockerfile on its own and run it; people are going to clone the repository. And we can document that that's how the build works for the tiny minority who won't do that.

Or if we really wanted to strictly adhere to the "only use things in the current directory" approach, couldn't we just have a `Dockerfile.dev` and `Dockerfile.prod` or something? Another reason I'm pushing to keep it in this repo is because it makes it easier to pull in new changes, especially if you patch things. It's the same reason I strongly recommend avoiding GitHub's tarball downloads and instead using a git clone (that is, if you really _must_ not use the npm package). It just makes things easier to keep track of.

adundovi replied:
OK. I'll try to put myself in the context of developing the code through the Docker flow to see how to adjust and adapt the Dockerfile to that need. Also, I'll try to find some examples of it.

Generally, Docker is not well designed from security stand points: you run it always with root or root-equivalent privileges which, for a dev environment at least, is a bad practice, right? When Dockerfile is copying and adding files from the outside of its own dir, it doesn't inspire confidence...

strugee replied:
Wait wait wait. I thought we decided to only focus on the production Docker image in this PR, and split the dev environment off into a separate issue?

> When Dockerfile is copying and adding files from the outside of its own dir, it doesn't inspire confidence...

I just don't see how this is a problem. Who cares about the difference between `.` and `../..`? If the Dockerfile was touching things outside the _repository clone_, say, `~/.local`, _that_ would be super sketchy. But I just don't see the argument that using things outside of the _directory_ is bad. Again, anyone using the Dockerfile will just be cloning the repository anyway. This feels a little like [cargo-cult programming][].

 [cargo-cult programming]: https://en.wikipedia.org/wiki/Cargo_cult_programming

adundovi replied:
It is not cargo-cult programming - it is a [documented](https://docs.docker.com/engine/reference/builder/#add) and intended behaviour if you don't trust my arguments:

> The <src> path must be inside the context of the build; you cannot ADD ../something /something, because the first step of a docker build is to send the context directory (and subdirectories) to the docker daemon.

Of course, there is a workaround, but I don't want to rape the procedure. Up to now, I have always used Docker for deployment, and I understand that context well, but, like I said, I will try to enter the mind of a "docker developer" to understand how to set it up properly and efficiently, probably by moving the dockerfile to the root of this repository as the first step.

strugee replied:
Ah, gotcha. I'm sorry for jumping to conclusions - from the way you were talking about it I just thought it was something you were "supposed" to do. I didn't realize there were actual technical restrictions; that's why I said "I just don't see the argument..." etc. Thanks for clarifying.

I think the best way forward is to just move the Dockerfile to the repo root, as you mentioned. Again, sorry for the confusion :)

strugee commented on deploy/docker/pump_wrapper.sh:
```diff
@@ -0,0 +1,97 @@
+#!/bin/sh
```
This is super hacky. We should just fix the underlying deficiency, which is #1202. I'll try to do that soon.

strugee replied:
#1202 is imminent: https://github.com/pump-io/pump.io/issues/1202#issuecomment-291050095

Please nuke this workaround :)

strugee commented:
Hey, thanks for working this! I'm going to close it since #1348 just got merged but I really appreciate it :)
